src/axes.jl

```diff
@@ -248,3 +247,64 @@ lazy_axes(x::AbstractRange, ::StaticInt{1}) = Base.axes1(x)
 lazy_axes(x, ::Colon) = LazyAxis{:}(x)
 lazy_axes(x, ::StaticInt{dim}) where {dim} = ndims(x) < dim ? SOneTo{1}() : LazyAxis{dim}(x)
 @inline lazy_axes(x, dims::Tuple) = map(Base.Fix1(lazy_axes, x), dims)
+
+"""
+    index_labels(x)
+    index_labels(x, dim)
+
+Returns a tuple of labels assigned to each axis or a collection of labels corresponding to
+axis `dim` of `x`. Default is to simply return `map(keys, axes(x))`.
+"""
+index_labels(x, dim)  = index_labels(x, to_dims(x, dim))
```

Comment:
Shouldn't the function live in ArrayInterfaceCore so that existing "named array" packages can overload it? BTW, it would be good to ping their authors to ensure they would all be OK with the API, otherwise it won't make a lot of sense.

Reply:
Existing packages can overload it from `ArrayInterface`. If you're referring to supporting named dimensions like NamedDims.jl, then they define `ArrayInterface.dimnames` and `to_dims` maps to the appropriate dimension so they don't have to overload every method with a dim argument.

Reply:
ArrayInterface is relatively heavy, which is why ArrayInterfaceCore was created IIUC. I guess it's up to package authors to say whether a dependency on ArrayInterface is acceptable for them or not.

Reply:
That was more of an issue before we went through all the trouble to fix invalidations due to `StaticInt` earlier this year. That doesn't mean we can't improve the situation. We are actively trying to move matured functionality into base where appropriate (see #340). I regularly review the code here in an effort to eliminate problematic code that still exists (e.g., redundancies, generated functions, etc,). For example, once we know how this PR is going to look I can finally finish an effort to consolidate a lot of "indexing.jl" by overloading base methods instead of reimplementing a lot of what's in base already.

Reply:
Also, it's a bit of a chicken and egg issue. We use `StaticSymbol` for dimension names known at compile time so that we can use them as a point of reference in an inferrible way. It's pretty difficult to do this only relying on constant propagation (demonstrated that with static sizes here https://github.com/JuliaLang/julia/pull/44538#issuecomment-1229205583).

If someone has a reliable solution I'm open to it. I've been trying to actively address this and related issues for years

Reply:
@oxinabox is it still a concern depending on ArrayInterface at this point?

Reply:
Now that ArrayInterface doesn't like Requires.jl a billion packages, I am much more comfortable depending upon it for NamedDims.jl

Reply:
My suggestion was just to put empty function definitions in ArrayInterfaceCore (like we do with DataAPI and StatsAPI). That doesn't prevent keeping fallback method definitions in ArrayInterface as this PR does. But packages that don't want to use fallback definitions are still able to overload the functions by depending only on ArrayInterfaceCore.

Reply:
We've been trying to avoid situations where the behavior of a method is different if ArrayInterface is loaded vs just ArrayInterfaceCore.